client/lib/site-spec/script-loader.ts

```diff
@@ -0,0 +1,182 @@
+	const config = RESOURCE_CONFIGS[ type ];
+	const selector =
+		type === 'script'
+			? `script[${ config.urlAttr }="${ url }"]`
+			: `link[${ config.urlAttr }="${ url }"]`;
+	return !! document.querySelector( selector );
+}
+
+/**
+ * Loading logic
+ */
+
+/**
+ * Generic injector used by both CSS & script.
+ * Creates and injects DOM elements for resource loading.
+ * @param url - The URL of the resource to inject
+ * @param type - The type of resource ({@link ResourceType})
+ * @returns Promise that resolves when the resource is loaded
+ * @throws {Error} When resource loading fails
+ */
+function injectResource( url: string, type: ResourceType ): Promise< void > {
+	ensureDomAvailable();
+	const config = RESOURCE_CONFIGS[ type ];
+
+	return new Promise< void >( ( resolve, reject ) => {
+		const element = document.createElement( config.tagName ) as HTMLScriptElement | HTMLLinkElement;
+
+		element.id = config.elementId;
+
+		// Set common attributes
+		for ( const [ attributeKey, attributeValue ] of Object.entries( config.attributes ) ) {
+			element.setAttribute( attributeKey, attributeValue );
+		}
+
+		// Set URL attribute based on resource type (src for script, href for CSS)
+		if ( config.urlAttr === 'src' ) {
+			const scriptElement = element as HTMLScriptElement;
+			scriptElement.src = url;
+			scriptElement.async = true;
+		} else {
+			const linkElement = element as HTMLLinkElement;
+			linkElement.href = url;
+		}
+
+		element.onload = () => resolve();
+		element.onerror = () => reject( new Error( `${ config.errorMessage } ${ url }` ) );
+
+		document.head.appendChild( element );
+	} );
+}
+
+/**
+ * Ensures a specific resource is loaded, with intelligent caching to prevent duplicate requests.
+ * @private
+ * @param url - The URL of the resource to load
+ * @param type - The type of resource ({@link ResourceType})
+ * @returns Promise that resolves when the resource is loaded
+ */
+function ensureResourceLoaded( url: string, type: ResourceType ): Promise< void > {
+	// Exists in DOM? done
+	if ( isResourceInDOM( url, type ) ) {
+		return Promise.resolve();
+	}
+
+	const key = cacheKeyFor( type, url );
+	const existingPromise = loadingPromises.get( key );
+	if ( existingPromise ) {
+		return existingPromise;
+	}
+
+	// Start a new load and cache it until it settles.
+	const loadPromise = injectResource( url, type ).finally( () => {
+		loadingPromises.delete( key );
+	} );
+
+	loadingPromises.set( key, loadPromise );
+	return loadPromise;
+}
+
+/**
+ * Loads a specific resource type (script or CSS).
+ * @param type - The type of resource to load ({@link ResourceType})
+ * @returns Promise that resolves when the resource is loaded
+ * @throws {Error} When resource URL is not configured
+ */
+async function loadResource( type: ResourceType ): Promise< void > {
+	const url = getSiteSpecUrlByType( type );
+	if ( ! url ) {
+		return Promise.reject( new Error( `SiteSpec ${ type } URL not configured` ) );
```

Comment:
v-minor: using `throw` is probably nicer.

```suggestion
		throw new Error( `SiteSpec ${ type } URL not configured` );
```